qiskit/dagcircuit/dagcanonical.py

```diff
@@ -0,0 +1,400 @@
+            yield node
+
+    def edges(self):
+        """Function to yield all edges.
+
+        Yields:
+            Iterator: generate all edges.
+        """
+        for edge in self._multi_graph.edges(data=True):
+            yield edge
+
+    def in_edge(self, node_id):
+        """ Get the list of incoming nodes for a given node_id.
+
+        Args:
+            node_id (int): id of the corresponding node
+
+        Returns:
+            list[In_edges()]: List of all incoming edges.
+        """
+        return self._multi_graph.in_edges(node_id)
+
+    def out_edge(self, node_id):
+        """List of all outgoing edges for the given node id.
+
+        Args:
+            node_id (int): id of the corresponding node.
+
+        Returns:
+            list[out_edges()]: List of all incoming edges.
+        """
+        return self._multi_graph.out_edges(node_id)
+
+    def direct_successors(self, node_id):
+        """Returns label of direct successors of a node as sorted list """
+        return sorted(list(self.to_networkx().successors(node_id)))
+
+    def direct_predecessors(self, node_id):
+        """Returns label of direct  predecessors of a node as sorted list """
+        return sorted(list(self.to_networkx().predecessors(node_id)))
+
+    def successors(self, node_id):
+        """Returns set of the ancestors of a node as DAGNodes."""
+        return self.to_networkx().nodes[node_id]['successors']
+
+    def predecessors(self, node_id):
+        """Returns set of the descendants of a node as DAGNodes."""
+        return self.to_networkx().nodes[node_id]['predecessors']
+
+    def draw(self, scale=0.7, filename=None, style='color', category='canonical'):
+        """
+        Draws the DAG canonical circuit.
+
+        This function needs `pydot <https://github.com/erocarrera/pydot>`, which in turn needs
+        Graphviz <https://www.graphviz.org/>` to be installed.
+
+        Args:
+            scale (float): scaling factor
+            filename (str): file path to save image to (format inferred from name)
+            style (str): 'plain': B&W graph
+                         'color' (default): color input/output/op nodes
+            category(str): 'canonical' Other type of DAG
+
+        Returns:
+            Ipython.display.Image: if in Jupyter notebook and not saving to file,
+                otherwise None.
+        """
+        from qiskit.visualization.dag_visualization import dag_drawer
+        return dag_drawer(dag=self, scale=scale, filename=filename, style=style, category=category)
+
+
+def merge_no_duplicates(*iterables):
+    """Merge K list without duplicate using python heapq ordered merging
+
+    Args:
+        *iterables: A list of k sorted lists
+
+    Yields:
+        Iteraor: List from the merging of the k ones (without duplicates
+    """
+    last = object()
+    for val in heapq.merge(*iterables):
+        if val != last:
+            last = val
+            yield val
+
+
+def _commute(node1, node2):
```

Comment:
this seems to be the same function as the one in `passes/commutation_analysis.py`, perhaps improved?

Actually the `CommutationAnalysis` pass should not be needed if we can somehow make use of this DAGCanonical rather than DAG.

Reply:
@ajavadia Thank you for the comments and happy new year!

I read more about the passes and transpilers and have now a better understanding of them. Then you are right that it would be better to use converters DAGCircuit <-> DAGCanonicalCircuit. First the main goal that I have (at the moment) for the canonical form is to implement a template matching pass: arXiv:1909.05270 In the current version, they are only converters form circuit to dag canonical and vice versa. Then I will update the code to introduce a converter from the DAGcircuit.

As you said it can be also useful to use this form for commutation related passes. The function for checking commutation is indeed an updated version of the one in CommutationAnalysis. I tried to introduce considerations on which qubits the operation are acting on. I also added a list (which is not complete) for known commutation relations. I also changed the commutation rules for measurement (qubit dependent). The function is still not optimal and any comments are welcome.

Reply:
Ok sounds good. Please add the DAGCircuit <-> DAGCanonical conversions, and @kdk and I will review.